drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt

```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
         Logger.d(TAG, "onPaymentMethodsResponseChanged")
+        setupStoredPaymentMethods(storedPaymentMethods)
+        setupPaymentMethods(paymentMethods)
+    }
+
+    private fun setupStoredPaymentMethods(storedPaymentMethods: List<StoredPaymentMethod>) {
+        storedPaymentMethodsList.clear()
+        for (storedPaymentMethod in storedPaymentMethods) {
+            val type = storedPaymentMethod.type
+            val id = storedPaymentMethod.id
+            if (type != null && id != null && PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type)) {
+                // We don't check for availability on stored payment methods
+                storedPaymentMethodsList.add(makeStoredModel(storedPaymentMethod))
+            } else {
+                Logger.e(TAG, "Unsupported stored payment method - $type - $id")
+            }
+        }
+    }
+
+    private fun makeStoredModel(storedPaymentMethod: StoredPaymentMethod): StoredPaymentMethodModel {
+        return when (storedPaymentMethod.type) {
+            PaymentMethodTypes.SCHEME -> {
+                StoredPaymentMethodModel(
+                    storedPaymentMethod.id ?: "",
+                    "•••• ${storedPaymentMethod.lastFour}",
+                    "${storedPaymentMethod.expiryMonth}/${DateUtils.removeFirstTwoDigitFromYear(storedPaymentMethod.expiryYear)}",
+                    storedPaymentMethod.brand ?: ""
+                )
+            }
+            else -> StoredPaymentMethodModel(
+                storedPaymentMethod.id ?: "",
+                storedPaymentMethod.name ?: "",
+                "",
+                storedPaymentMethod.type ?: ""
+            )
+        }
+    }
+
+    private fun setupPaymentMethods(paymentMethods: List<PaymentMethod>) {
+        // We can't remove availability callbacks so just for safety let's crash in case of a concurrency issue.
+        if (availabilityChecksum != 0) {
+            throw CheckoutException("Concurrency error. Cannot update Payment methods list because availability is still being checked.")
+        }
+
+        // Reset variables
+        availabilitySum = 0
+        availabilitySkipSum = 0
+        availabilityChecksum = paymentMethods.size
+        paymentMethodsList.clear()
 
         for (paymentMethod in paymentMethods) {
             val type = paymentMethod.type
+            when {
+                type == null -> {
+                    throw CheckoutException("PaymentMethod type is null")
+                }
+                PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.v(TAG, "Supported payment method: $type")
+                    // We assume payment method is available and remove it later when the callback comes
+                    // this is the overwhelming majority of cases, and we keep the list ordered this way.
+                    paymentMethodsList.add(
+                        PaymentMethodModel(type, paymentMethod.name ?: nullNameFallback)
```

Comment:
What about localization here? The UI will show "Unknown" for all languages. Or is this an edge case that shouldn't practically happen?

Reply:
Good point, this shouldn't really happen, I was also considering to just leave "" for the fallback. Maybe that's better and the View can have the localized "Unknown".